src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunTableManager.cs

```diff
@@ -53,12 +55,22 @@ protected virtual void Graph_NewMarkedNode(DependencyNodeCore<NodeFactory> obj)
 
             if (methodNode != null)
             {
-                _methodsGenerated.Add(methodNode.Method);
+                lock (_methodsGenerated)
+                {
+                    _methodsGenerated.Add(methodNode.Method);
+                }
             }
         }
         public IEnumerable<MethodDesc> GetCompiledMethods()
         {
-            return _methodsGenerated;
+            lock (this)
+            {
+                if (_sortedMethods == null)
+                {
+                    _sortedMethods = new List<MethodDesc>(_methodsGenerated.OrderBy(method => method, new TypeSystemComparer()));
```

Comment:
`GetCompiledMethods` *should* be called after all nodes are added to the graph. Currently we rely on all nodes calling it early out for `relocsOnly` so they only call it when we're emitting the binary.  If we ever mess that up with a new node, we might generate `_sortedMethods` and then any calls to `Graph_NewMarkedNode` after that point will add to `_methodsGenerated` but we won't generate `_sortedMethods` more than once (so we'd silently be missing methods).  I would recommend a couple things here:
1. After we sort the methods, null out `_methodsGenerated` since we no longer need that duplicated data. It'll get GC'd and claw back some memory.
2. In `Graph_NewMarkedNode`, assert that `_methodsGenerated != null` as a helper to devs that fall into this trap in the future.

Reply:
Done.